### DISCUSSION TITLE
Index with Map and Reduce that Looks at Internal Types

### DISCUSSION THREAD

**Discussion by mfonnesbeck** [ravendb/ravendb#17070] [Q&A]
I have an object that contains a history array.  The base type of history items is TicketHistoryItem which contains a Timestamp value of when the history change was made, but this history item in the array could be either of the following descendant classes: TicketStatusHistoryItem holding a change in statuses of the original object, a TicketCategoryHistoryItem holding changes in the category of the original object, or a two other types of changes. Internal in the stored document it has a $type value storing the actual class this history item is and lists the fully qualified namespace and class name.

There is a need for querying these objects and grouping them by the TicketStatusHistoryItem history types and getting a count of them based upon the dates (Timestamp) that the changes were made. In order to count how many were a particular status on a date. 

Is there a way to create an index that looks at the $type property and filters those objects that contain a TicketStatusHistoryItem with a particular status in the history array? I've tried various ways of accessing this array and filtering/grouping without success and not found any examples as complicated as this in the docs.

**Comment by ayende** [comment]
Remember that on the server side, you don't actually have your types. It is all JSON.

What you can do is use a JS index, and just operate over the json structure directly
